### PR TITLE
[Canonical structures] Annotation to field declarations to prevent them from being “canonical”

### LINKED ISSUE #4796: Disable canonical projections
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4796
From: @robbertkrebbers
Reported version: 8.5
CC:   @maximedenes, @RalfJung

### PR DESCRIPTION
**Pull request by vbgl**
Depends on #10075. (merged)
Fixes #4796.

 - [x] Needs documentation.
 - [x] Needs overlay for elpi (https://github.com/LPCIC/coq-elpi/pull/57)

### CODE REVIEW

File: pretyping/recordops.mli
Comment (SkySkimmer) [inline pretyping/recordops.mli]: Maybe label the bool arguments so we can differentiate them.
Or remove the `mk_proj_kind` so people have to use the record building syntax `{pk_name = ...; ...}`.

**Comment by vbgl** [comment]
@gares Suggestion for the user manual welcome. Thanks!

File: doc/changelog/02-specification-language/10076-not-canonical-projection.rst
Comment (Zimmi48) [inline doc/changelog/02-specification-language/10076-not-canonical-projection.rst]:
```suggestion
  (`#10076 <https://github.com/coq/coq/pull/10076>`_,
```

Sorry, rst syntax is sometimes quite weird.

**Review by Zimmi48 — COMMENTED** [review]
I don't like this: you are reusing the syntax of attributes within a term without making the support generic. There are many cases where having attributes within a term would be useful, so if you really want to reuse this syntax, make it generic!

File: doc/changelog/02-specification-language/10076-not-canonical-projection.rst
Comment (vbgl) [reply doc/changelog/02-specification-language/10076-not-canonical-projection.rst]: Can you please explain what it does change? Thank you.

File: doc/sphinx/addendum/canonical-structures.rst
Comment (gares) [inline doc/sphinx/addendum/canonical-structures.rst]: What bad happens? I'm a bit lost ;-)

File: vernac/vernacexpr.ml
Comment (gares) [inline vernac/vernacexpr.ml]: can't we just have a bool here, and maybe remove the negation (having `true` as default)?

File: doc/sphinx/addendum/canonical-structures.rst
Comment (vbgl) [reply doc/sphinx/addendum/canonical-structures.rst]: See the warning at the end of that page of the manual: https://coq.github.io/doc/master/refman/addendum/canonical-structures.html

**Comment by gares** [comment]
As usual syntax is debatable but I'd prefer `#[canonical(false)]` or `#[canonical = false]` (both match our syntax of attributes AFAICT).

File: doc/sphinx/addendum/canonical-structures.rst
Comment (gares) [reply doc/sphinx/addendum/canonical-structures.rst]: I looked at the bug report and @robbertkrebbers does not say why he wants to disable CS on a projection. I've nothing against it, but (apart from not polluting indexes or the output of Print Cano..) I don't know why one would want to disable CS.

File: doc/changelog/02-specification-language/10076-not-canonical-projection.rst
Comment (Zimmi48) [reply doc/changelog/02-specification-language/10076-not-canonical-projection.rst]: See the difference between the entry you added and the other at https://coq.gitlab.io/-/coq/-/jobs/209600067/artifacts/_install_ci/share/doc/coq/sphinx/html/changes.html.

Comment (vbgl) [reply doc/changelog/02-specification-language/10076-not-canonical-projection.rst]: Spotted! Thanks.

**Comment by Zimmi48** [comment]
> As usual syntax is debatable but I'd prefer `#[canonical(false)]` or `#[canonical = false]` (both match our syntax of attributes AFAICT).

Just to be clear. I'm not starting a discussion about syntax. I'm claiming that reusing the attribute syntax should only be allowed if the mechanism is generic, like attributes.

File: doc/sphinx/addendum/canonical-structures.rst
Comment (gares) [reply doc/sphinx/addendum/canonical-structures.rst]: Ah ok, here I'd say "it plays no role in the search of instances".

**Comment by gares** [comment]
> if the mechanism is generic, like attributes.

To me attributes have a generic syntax, and some of them can apply to many different statements. But I don't get where you are drawing the line, exactly. For example I'd like to write `#[canonical] Definition nat_eqType`, do you see any problem with that?

File: doc/sphinx/addendum/canonical-structures.rst
Comment (maximedenes) [reply doc/sphinx/addendum/canonical-structures.rst]: What about the warning?

**Comment by gares** [comment]
Oh, maybe I get it. Are you saying that the grammar rule should not be had-hoc for `canonical`?

**Comment by vbgl** [comment]
I’d like to disagree: we can make the handling of attributes generic later, as a completely unrelated change.

Reusing (mimicking) the syntax of attributes has the advantage of not inventing yet-another-ad-hoc-creative syntax.

File: doc/sphinx/addendum/canonical-structures.rst
Comment (vbgl) [reply doc/sphinx/addendum/canonical-structures.rst]: Indeed, the warning should be documented. I find it hard to understand.

File: vernac/vernacexpr.ml
Comment (vbgl) [reply vernac/vernacexpr.ml]: Done.

**Comment by Zimmi48** [comment]
> Oh, maybe I get it. Are you saying that the grammar rule should not be had-hoc for `canonical`?

Yes, that's what I'm saying.



> I’d like to disagree: we can make the handling of attributes generic later, as a completely unrelated change.

Certainly not unrelated.

> Reusing (mimicking) the syntax of attributes has the advantage of not inventing yet-another-ad-hoc-creative syntax.

It makes sense only if there is a clear plan and if this syntax is not going to stand in the way of this plan.

**Comment by gares** [comment]
> Yes, that's what I'm saying.

So you are asking @vgbl to just re-use the already existing nonterminal and filter a posteriori non supported attributes. I think it is reasonable to do so. At the same time, I don't see which discussion this PR needs (you added a label).

**Comment by Zimmi48** [comment]
Discussion would be needed only if @vbgl doesn't want to follow this suggestion. Otherwise, the label can be removed.

**Comment by maximedenes** [comment]
> this suggestion

What suggestion?

**Comment by gares** [comment]
> So you are asking @vbgl to just re-use the already existing nonterminal and filter a posteriori non supported attributes. I think it is reasonable to do so.

I guess this one

File: doc/sphinx/addendum/canonical-structures.rst
Comment (robbertkrebbers) [reply doc/sphinx/addendum/canonical-structures.rst]: > apart from not polluting indexes or the output of Print Cano..

That was indeed the main rationale.

**Comment by vbgl** [comment]
It now makes use of the generic parsing rules for attributes.

Should I extend the rules for boolean attributes to allow the `canonical(false)` variant proposed by Enrico? (In fact all four of `canonical(true)`, `canonical(false)`, `not_canonical(true)`, and `not_canonical(false)`)

**Comment by vbgl** [comment]
I got rid of `not_canonical`: you may now use `canonical(false)`.

As a side-effect, `noprogram` is gone too.

NB: `canonical=false` is not valid syntax. I may add support for `canonical="false"` if needs be.

**Review by Zimmi48 — COMMENTED** [review]
Thanks! This looks good.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I'm 2nd maintainer for pretyping, I can take this, but I'd like an hack from someone on the vernac business.

**Comment by Zimmi48** [comment]
@vbgl The current documentation for canonical structures is already dispersed: the `Record` command definition in the Gallina extensions chapter (a very messy first section), the canonical structure sub-section of the implicit arguments section in the Gallina extensions chapter, the Structure command definition in the Implicit coercions chapter, the examples in the Canonical structures chapter... so whatever you decide to document it, even with an example, will be better than nothing until we can solve this mess.

**Comment by herbelin** [comment]
> Reusing (mimicking) the syntax of attributes has the advantage of not inventing yet-another-ad-hoc-creative syntax.

Strictly speaking, to avoid "yet-another-ad-hoc-creative syntax", you could reuse existing forms of syntax, e.g.:
```
Structure type := _Pack { obj : Type; class_of : class obj } (class_of inactive in canonical structures).
Structure type := _Pack { obj : Type; non canonical class_of : class obj }.
```
On the other side, you also have been creative yourself! For instance, you made a choice in where to exactly write the attribute, you made a choice in the name(s) of the attribute, in the format of the attribute (parentheses vs =)! These choices are not anecdotal in themselves.

But I agree that choosing a "good" syntax is very difficult (good in the sense of smoothly expressing what it means, not too cryptic, not too verbose, easy to remember, easy to use, spontaneously talking to a lot of people...).

Otherwise, on the more technical side, would it alternatively make sense to look for a criterion able to "decide" automatically when a projection "should" not be declared as canonical projection? I.e., do we understand well the generality of the situations covered by #4796?

**Comment by SkySkimmer** [comment]
>`Structure type := _Pack { obj : Type; non canonical class_of : class obj }.`

That doesn't really work as it's parsed to something like `Structure type := _Pack { obj : Type; non : forall canonical class_of, class obj }.`

**Comment by herbelin** [comment]
> That doesn't really work as it's parsed to something like `Structure type := _Pack { obj : Type; non : forall canonical class_of, class obj }`.

You're right. I did not pay attention that it overlaps with an existing syntax (which it would then hide in this particular case). This is thus not so nice. (An alternative in the same spirit would however be to use e.g. `non-canonical`.)

But, anyway, I'm not willing to engage in a debate about a particular syntax or not. That's true that I generally prefer that such question is treated from a factual point of view, putting the spectrum of alternatives into perspective, so that we are each of us individually able to compare the advantage of this or that choice. That's true that I generally prefer that we are able to say to users: look, this has been discussed in length, pondered in length, and this is our collective choice. But more than that, we must also decide, and eventually trust and be supportive of each other choices, and in particular here of the choices of @vbgl who is investing a lot of time in addressing user needs.

**Comment by vbgl** [comment]
Thank you Hugo for your support.

About the position of the annotation, I indeed *chose* to put it before the field declaration (as in [Rust](https://doc.rust-lang.org/reference/items/structs.html)) rather than after it (as in [OCaml](https://caml.inria.fr/pub/docs/manual-ocaml-4.07/extn.html#sec260)). Is would be easy to allow both forms (so that each writer can choose according to their liking) with the same interpretation (so that the readers are not confused).

Moreover, the initial motivation to this change is, beyond satisfying Robbert’s wishes, is to be able to write code as described in the manual in real developments: named fields without precluding later declarations of canonical instances.

**Comment by gares** [comment]
Testsuite failure seems spurious, I've restarted it. Mtac2 one is unrelated.

There is still a `needs: documentation` label to be addressed. Then we are good to go I guess.

**Comment by herbelin** [comment]
> Moreover, the initial motivation to this change is, beyond satisfying Robbert’s wishes, is to be able to write code as described in the manual in real developments: named fields without precluding later declarations of canonical instances.

Actually, there is something that I don't understand. Why to address the issue at the level of `Record` rather than at the level of `Canonical Structure`. When we define a record, the purpose is not necessarily to use it at once with canonical structures. Moreover, a `Record` statement can be given only once. What to do if one wants to change the status of the projection after a `Record` declaration has been given?

**Comment by vbgl** [comment]
This is exactly the point.

The `Canonical` command can be used to turn a constant that plays no role in the inference of canonical instances into something that plays a role.

Currently, defining a record automatically turns all named fields (that are not local definitions, aka let-in) into something that plays a role. There is no way to disable this behavior.

This PR adds a way for the user to declare: “here is a named field, it should not mess up with instance resolution”.

**Comment by Zimmi48** [comment]
This makes me wonder if the default should not be off. WDYT?

File: doc/sphinx/language/gallina-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-extensions.rst]: Please add an example here, because it is quite hard to understand this sentence out of context. In particular, the declaration of a field is a bit weird conceptually: it is the record which is declared.

**Comment by vbgl** [comment]
Making « false » the default would break quite a large code base, I fear.

**Comment by ggonthier** [comment]
I'll second @herbelin 's suggestion of providing per-instance control. There are quite a few `Canonical` declarations in MathComp that only intend to declare canonical values for _some_ of the projectors of a structure, sometimes resorting to the `unkeyed` macro to disable registration. This makes the `projection-no-head-constant` and `redundant-canonical-projection` warnings too spurious to be useful and we have to turn them off.

**Comment by gares** [comment]
I agree, but given this PR is ready I'd rather merge it and leave the per instance annotation for another PR.
See #10140

**Comment by vbgl** [comment]
Thank you Georges for the feed-back. Do you think that per-instance tuning would be enough and render the per-structure (implemented in this PR) useless?

For instance, one of the issues I’m attempting to solve is the warnings that show up in the [manual](https://coq.github.io/doc/master/refman/addendum/canonical-structures.html):

> Warning: Ignoring canonical projection to LEQ.Class by LEQ.class_of in pair_LEQty: redundant with nat_LEQty

In this particular example, should I annotate both instance declarations (`nat_LEQty` and `pair_LEQty`)?

**Comment by gares** [comment]
> Making « false » the default would break quite a large code base, I fear.

IIRC in mathcomp se systematically use `Structure` instead of `Record` for things to be used with CS inference. Maybe this could be a way to change the default value of the attribute.

**Comment by Zimmi48** [comment]
Then `Record` would mean with a default of `false` and `Structure` would mean with a default of `true`. That's a good idea, I think. Obviously, as usual, there could be global flags to control that.

**Comment by vbgl** [comment]
Ping @gares

**Comment by herbelin** [comment]
Sorry to come back to this. I'm not a specialist of canonical structures (and in particular not aware of the exact examples involved), but were the concerns expressed by @ggonthier and I addressed?

**Comment by vbgl** [comment]
You are right. This PR is just a starting point. There is some future work to be done, in particular enabling per-instance control of the relevant projectors.

**Comment by herbelin** [comment]
> There is some future work to be done, in particular enabling per-instance control of the relevant projectors.

But then, I'm missing something. What is the point of providing an irreversible control at `Structure` declaration time, if you also independently provide a more general reversible way to control the relevant projectors?

**Comment by vbgl** [comment]
At structure-declaration time, you can choose what you think is the good default for that structure. At instance-declaration time, you can selectively override that default.

**Comment by herbelin** [comment]
> At structure-declaration time, you can choose what you think is the good
default for that structure. At instance-declaration time, you can selectively
override that default.

Yes, why not. Then, it is probably worth that other people express an opinion on the different strategies that can be followed. Maybe I missed offline discussions but I feel that letting more time for a constructive discussion would be very welcome in such situation.

Also, what implementation do you have in mind for selectively overriding a default at instance-declaration time?

**Comment by Zimmi48** [comment]
I don't really feel concerned by this discussion but about the “letting more time for a constructive discussion”, it seems to me that the long time that we still have before the 8.11 release should help refine the design here if needed. The fact that it is after merge of a preliminary step doesn't change much. And if such discussion didn't happen (but I hope that it does given that #10140 in particular has been opened), it seems to me that this first step would still be better than nothing, wouldn't it?

>Also, what implementation do you have in mind for selectively overriding a default at instance-declaration time?

I'd suggest you ask this question in #10140, otherwise the information risk being scattered...

**Comment by herbelin** [comment]
> I'd suggest you ask this question in #10140, otherwise the information risk being scattered...

Ah sorry, I did not see that the discussion was continuing in #10140. Then, I would also tend to explore the feasability of an `Attributes` command which would allow to amend/revert a registration made by `#[canonical(false)]` or `Canonical Structure`/`Canonical`. Or maybe even some `Unification Hint` directive.

**Pull request merged**
